### ISSUE TITLE
CPU training stalls for 60s+ compiling native artifact for layer/conv contraction kernels with integer quantization

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#357]
Earliest public failure: one epoch of recipe.train().run() on backend cpu remains active past 60 seconds for recipe.data("data/temporal/series_npz.npz").target("target").norm(z_score) with

```rust
let model = recipe.model()
	.layer(8).sigmoid().qi(5).k
	.conv(8, 1).elu().norm(batch).iq(2).s
	.loss(ce);
```

Inspected implementation cause: train preparation, where compile_model emits the graph IR and compile_native_artifact invokes the external CPU compiler (clang -O2 -x ir) on it. Workload check: about 1.5e5 loaded rows, a handful of features, a few hundred parameters, under 0.2 GFLOP of forward, backward, and optimizer work, and far under 0.1 GB of memory traffic justify an epoch in low single-digit seconds even single threaded, so epoch execution cannot supply the gap. The IR heaviness comes from the contraction kernels of both blocks plus inline qi(5).k and iq(2).s weight decode. Closed #167 recorded this same compile_native_artifact stall for conv contraction kernels; this stable replay at the current base commit shows that path stalling again and no open issue covers it.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `fb66e0f8c091a1b7`
- Classifier provider: `opencode`
- Classifier model: `x-preview-f-free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "new",
  "issue": 0,
  "title": "CPU training stalls for 60s+ compiling native artifact for layer/conv contraction kernels with integer quantization",
  "body": "Earliest public failure: one epoch of recipe.train().run() on backend cpu remains active past 60 seconds for recipe.data(\"data/temporal/series_npz.npz\").target(\"target\").norm(z_score) with\n\n```rust\nlet model = recipe.model()\n\t.layer(8).sigmoid().qi(5).k\n\t.conv(8, 1).elu().norm(batch).iq(2).s\n\t.loss(ce);\n```\n\nInspected implementation cause: train preparation, where compile_model emits the graph IR and compile_native_artifact invokes the external CPU compiler (clang -O2 -x ir) on it. Workload check: about 1.5e5 loaded rows, a handful of features, a few hundred parameters, under 0.2 GFLOP of forward, backward, and optimizer work, and far under 0.1 GB of memory traffic justify an epoch in low single-digit seconds even single threaded, so epoch execution cannot supply the gap. The IR heaviness comes from the contraction kernels of both blocks plus inline qi(5).k and iq(2).s weight decode. Closed #167 recorded this same compile_native_artifact stall for conv contraction kernels; this stable replay at the current base commit shows that path stalling again and no open issue covers it.",
  "rationale": "Searched the tracker for slow runtime, CPU 60s stalls, conv/contraction compilation, quantized decoding, and npz datasets, then read the plausible matches: #164, #167 (closed), #250, #253, #278, #279, #287, #290, plus title scans of the scan/predictor/nv0 families. Open issues cover other lowering paths (scan, attention/pool, pool/residual with batch norm and quantized activation, predictor kernels) or execution phases (kmeans, MoE routing, CatBoost fitting), none covers plain layer/conv contraction artifact compilation. Source inspection: decode_tables npz loader and prepare are linear over about 1.5e5 rows; lower_project and lower_conv build two small Contraction nodes; NativeTape::epoch is one native launch whose arithmetic and memory traffic justify low single-digit seconds; the remaining consumer of the measured minute is compile_model writing IR and compile_native_artifact running clang -O2 -x ir (recipe.rs 2828 to 2904, 6353). #167 is the closed Conv/Contraction counterpart, so a fresh stable reproduction at commit 5306c211 through the identical path is not represented by any open issue and warrants a new report rather than a comment."
}
```

## Deterministic failure packet

kind=performance
id=fb66e0f8c091a1b7
base=commit=5306c211d9fc2b0186a4eaf6d9e25a10caec49ca tracked_tree=modified
cursor=seed:0 cursor:10367 next:10368 composition:100024220729293
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:cpu
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /tmp/recipe-composition-repro-cpu-10367-62.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/series_npz.npz")
		.target("target")
		.norm(z_score);
	let model = recipe.model()
		.layer(8).sigmoid().qi(5).k
		.conv(8, 1).elu().norm(batch).iq(2).s
		.loss(ce);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(100024220729293)
		.epochs(1)
		.log(all)
		.tf(32)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=cpu

**Comment by nm-z** [comment]
Closing as resolved based on zero-change verification PR #446.

That PR changes 0 files and verifies the existing `minimal` behavior, so there is no code to merge. The original verification is copied below.

<details>
<summary>Verification from PR #446: Verify current CPU layer convolution execution</summary>

Closes #357

## Verification

- Ran the exact reported temporal NPZ data, layer plus convolution graph, quantization, TF32 arithmetic, optimizer, seed, normalization, and one-epoch program through the current release CLI. Only the bundle path moved from `/tmp` to the disk-backed issue cache.
- Used the current CPU selection boundary, `RECIPE_FORCE_CPU=1` with `--device cpu`. The loader prepared 10,000 rows.
- The main TF32 BCE artifact was a cold cache miss. It compiled with Clang 22.1.8 for `x86_64-unknown-linux-gnu` and loaded as a native shared object; the 64-row calibration artifact was a cache hit.
- The run exited 0 in 0.73 seconds. Epoch 1 took 39.320 ms, BCE was finite at 0.6932, and the saved bundle was 227,276 bytes.
- The reported CPU contraction native-artifact compilation stall does not reproduce on current `minimal`.

## Change

This verification commit changes no tracked files because current behavior does not justify a production-code change.


</details>

Accepted verdict: PASS

Reason: The verification above reports that current `minimal` completes this issue's reported path successfully.

**Comment by nm-z** [comment]
The public `recipe.train().run()` remains active beyond 60 seconds on CPU while compiling the first quantized layer contraction through `NativeTape::new` -> `compile_model` -> `compile_native_artifact` using the CPU LLVM path. The LightGBM fit is downstream and does not change the earliest failure.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `690fef4198caa318`
- Classifier provider: `copilot`
- Classifier model: `gpt-5.6-luna`
- Classifier effort: `auto`

### Classifier decision

```json
{
  "verdict": "comment",
  "issue": 357,
  "title": "CPU training stalls for 60s+ compiling native artifact for layer/conv contraction kernels with integer quantization",
  "body": "The public `recipe.train().run()` remains active beyond 60 seconds on CPU while compiling the first quantized layer contraction through `NativeTape::new` -> `compile_model` -> `compile_native_artifact` using the CPU LLVM path. The LightGBM fit is downstream and does not change the earliest failure.",
  "rationale": "The loader produces about 12,000 rows with one source feature, represented as roughly 20 scalar text values. The first `.layer(4)` performs only about 12,000 * 20 * 4 multiply-accumulates per forward or epoch pass, with a few megabytes of memory traffic. The downstream LightGBM has 8 boosting iterations, at most 8 leaves, depth about 3, 16 histogram bins, and 4 post-layer features, requiring only several million row-feature scans and negligible memory. This workload should complete in well under 1 to a few seconds on the available CPU, so 60 seconds is materially unjustified and dominated by native LLVM compilation. Source inspection shows `lower_block` creates the quantized `Primitive::Contraction`, `lower_estimator` first calls `graph_inputs`, and `NativeTape::new` compiles that partial graph before LightGBM fitting. Issue #357 represents this same earliest CPU native-artifact compilation path for integer-quantized layer/conv contractions; #277 and the LightGBM-specific issues are downstream or backend-mismatched."
}
```

## Deterministic failure packet

kind=performance
id=690fef4198caa318
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=modified
cursor=seed:0 cursor:24419 next:24420 composition:116772571166521
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:cpu
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-9367-24419/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/chronological_splits")
		.target("target")
		.norm(z_score);
	let model = recipe.model()
		.layer(4).ln().iq(2).m
		.lgbm().prelu().qi(4).nf
		.loss(huber);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(116772571166521)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.int(4)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=cpu
